src/haz3lcore/dynamics/TypeAssignment.re

```diff
@@ -0,0 +1,288 @@
+    Typ.eq(ty, Bool) ? Some(Typ.Prod([])) : None;
+  | BoolLit(_) => Some(Bool)
+  | IntLit(_) => Some(Int)
+  | FloatLit(_) => Some(Float)
+  | StringLit(_) => Some(String)
+  | BinBoolOp(_, d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    Typ.eq(ty1, Bool) && Typ.eq(ty2, Bool) ? Some(Typ.Bool) : None;
+  | BinIntOp(op, d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    if (Typ.eq(ty1, Int) && Typ.eq(ty2, Int)) {
+      switch (op) {
+      | Minus
+      | Plus
+      | Times
+      | Power
+      | Divide => Some(Typ.Int)
+      | LessThan
+      | LessThanOrEqual
+      | GreaterThan
+      | GreaterThanOrEqual
+      | Equals
+      | NotEquals => Some(Typ.Bool)
+      };
+    } else {
+      None;
+    };
+  | BinFloatOp(op, d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    if (Typ.eq(ty1, Float) && Typ.eq(ty2, Float)) {
+      switch (op) {
+      | Minus
+      | Plus
+      | Times
+      | Power
+      | Divide => Some(Typ.Float)
+      | LessThan
+      | LessThanOrEqual
+      | GreaterThan
+      | GreaterThanOrEqual
+      | Equals
+      | NotEquals => Some(Typ.Bool)
+      };
+    } else {
+      None;
+    };
+  | BinStringOp(op, d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    if (Typ.eq(ty1, String) && Typ.eq(ty2, String)) {
+      switch (op) {
+      | Concat => Some(Typ.String)
+      | Equals => Some(Typ.Bool)
+      };
+    } else {
+      None;
+    };
+  | ListLit(_, _, ty, _) => Some(List(ty))
+  | Cons(d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    switch (ty2) {
+    | List(Unknown(Internal)) => Some(Typ.List(ty1))
+    | List(ty3) when Typ.eq(ty3, ty1) => Some(ty2)
+    | _ => None
+    };
+  | ListConcat(d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    switch (ty1, ty2) {
+    | (List(Unknown(Internal)), _)
+    | (_, List(Unknown(Internal))) => Some(Typ.List(Unknown(Internal)))
+    | (List(ty1), List(ty2)) when Typ.eq(ty1, ty2) => Some(Typ.List(ty1))
+    | _ => None
+    };
+  | Tuple(dhs) =>
+    let+ typ_list =
+      dhs |> List.map(typ_of_dhexp(ctx, m)) |> OptUtil.sequence;
+    Typ.Prod(typ_list);
+  | Prj(dh, i) =>
+    let* ty = typ_of_dhexp(ctx, m, dh);
+    switch (ty) {
+    | Prod(l) when List.length(l) != 0 => Some(List.nth(l, i))
+    | _ => None
+    };
+  | Constructor(_, typ) => Some(typ)
+  | ConsistentCase(Case(d_scrut, d_rules, _)) =>
+    let* ty' = typ_of_dhexp(ctx, m, d_scrut);
+    let* typ_cases: list(Typ.t) =
+      d_rules
+      |> List.map(rule_prj)
+      |> List.map(((dhp, de)) => {
+           typ_of_dhexp(dhpat_extend_ctx(dhp, ty', ctx), m, de)
+         })
+      |> OptUtil.sequence;
+    Typ.join_all(~empty=Unknown(Internal), ctx, typ_cases);
```

Comment:
we don't want to do a join here -- they should all be equal types

Reply:
@cyrus- What should happen if the types are not equal like in `if true then  else 1.` because the first case is Unknown(Internal) but the second case is Float?

Reply:
it should elaborate with a cast to the join type on the branches I believe